Grackle with COMOVING

YuriOku opened the pull request:
In this PR, I have modified the code to use Grackle with `COMOVING`. The changes I made are

- Define Grackle units in `Grackle_Init.cpp` and `Grackle_Prepare.cpp`
- Convert the specific thermal energy between GAMER and Grackle units in `Grackle_Prepare.cpp` and `Grackle_Close.cpp`
- Convert the timestep size to the physical time interval in `Grackle_AdvanceDt.cpp`
- Disable normalization of passive scalars for electron and deuterium densities for consistency with Grackle in `Init_Field.cpp`

I then created a test problem, `Grackle_Comoving`. The test simulation evolves uniform density gas with an initial temperature of T=1e8 K from z=9 to z=0. The time evolution of density and temperature is recorded in `Record__User`, and the cooling curve can be computed from the record with the Python script `Compute_CoolingCurve.py`. There is a C program `coolingrate_proper.c`, which generates a cooling curve table using Grackle at each redshift, which can be compared with the result of the test simulation.

In the left panel in the figure below, the solid line shows the cooling curve obtained from the test simulation, and the points indicate the values at each redshift. The dotted lines are cooling curves at each redshift. The points lie just on the dotted lines, indicating that the thermal evolution is consistent with the Grackle cooling curve.

![CoolingCurve](https://github.com/user-attachments/assets/140dc7e4-cbd7-4d50-bdec-7d64ac8f5c8a)

hyschive commented:
@YuriOku Thanks for the contribution! @ChunYen-Chen will help review your PR.

ChunYen-Chen commented:
Hi @YuriOku, I will review your PR this week and probably finish the review before the weekend.

ChunYen-Chen commented:
Hi @YuriOku, can you share the grackle data `CloudyData_noUVB.h5` with me?

YuriOku commented:
Hi Chun-Yen,

Please find the data file at
https://github.com/grackle-project/grackle_data_files/tree/main/input.

Thanks!

On Tue, Sep 10, 2024 at 11:08 AM Chun-Yen Chen ***@***.***>
wrote:

> Hi @YuriOku <https://github.com/YuriOku>, can you share the grackle data
> CloudyData_noUVB.h5 with me?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/gamer-project/gamer/pull/349#issuecomment-2339518989>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AO55JJFO5D4IT7KZJ4XIKKDZVZPBDAVCNFSM6AAAAABNZ4SOJCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGMZZGUYTQOJYHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

ChunYen-Chen reviewed (changes requested):
Hi @YuriOku,

The PR looks great and I can reproduce your results. I have two questions about your analysis script/tool:
1. Is there any specific reason you choose `-O2` to compile `coolingrate_proper.c`?
2. Can you run `coolingrate` with single precision grackle? If not, there should be some bugs in `coolingrate.c`.
   * If your tool does not support single precision, please add a check of `FLOAT8` in `Init_TestProb_Grackle_Comoving.cpp`

Before merging, please fix/change the following:
* Please fix all bugs with comments starting with [BUG].
* Please update your new code to GAMER code style with the comments starting with [Style]. You can resolve most of the code-style comments by copying the code I wrote in the comment.
* Please remove all the trailing spaces. You can search them by command:
   ```bash
   grep -r -n "[[:blank:]]$" ./
   ```

The rest of the comments without tags are some minor change suggestions or requests.  

Please let me know if any of my comments is unclear.

YuriOku commented:
Hi @ChunYen-Chen, thank you for your review! I will address your points.

> Is there any specific reason you choose `-O2` to compile `coolingrate_proper.c`?

There is no reason for the option.

> Can you run `coolingrate` with single precision grackle? If not, there should be some bugs in `coolingrate.c`.
> If your tool does not support single precision, please add a check of `FLOAT8` in `Init_TestProb_Grackle_Comoving.cpp`

I have error in running `coolingrate` with single precision grackle. I will add a check of `FLOAT8` in `Init_TestProb_Grackle_Comoving.cpp`

ChunYen-Chen commented:
Hi @YuriOku,
> > Is there any specific reason you choose -O2 to compile coolingrate_proper.c?

> There is no reason for the option.

How about changing it to `-O3`?

> > Can you run `coolingrate` with single precision grackle? If not, there should be some bugs in `coolingrate.c`.
> > If your tool does not support single precision, please add a check of `FLOAT8` in `Init_TestProb_Grackle_Comoving.cpp`

> I have error in running `coolingrate` with single precision grackle. I will add a check of `FLOAT8` in `Init_TestProb_Grackle_Comoving.cpp`

Great, please also update the `README` in your test problem.

YuriOku commented:
Hi @hyschive,

I see. I will try to run the code with a single precision. 

The floating-point precision of Grackle can be checked by `sizeof(gr_float)`.

YuriOku commented:
@hyschive @ChunYen-Chen 
I have tried to my code in single-float precision, but I could not make it. The [Grackle's example C code](https://github.com/grackle-project/grackle/blob/main/src/example/c_example.c) also fails in single-float precision. I could not identify the cause, but I think this is a bug in Grackle. 

Given that Grackle is not recommended run in single-float precision ([Grackle Installation Doc](https://grackle.readthedocs.io/en/latest/Installation.html)), I think it would be better to run Grackle in double precision regardless of the precision of fluid solver. I propose to define `h_Che_Array` in `gr_float`, and then, the type conversion is automatically made in `Grackle_Prepare` and `Grackle_Close`. An option like `GRACKLE_FLOAT` can be added to check the precision of Grackle. How do you think?

hyschive commented:
@YuriOku
- I agree with your suggestion and have just filed a feature request #354.
- On the other hand, IIUC, @ChunYen-Chen already performed the `Grackle_Comoving` test in both single and double precision and the results are consistent. So I guess the problem lies in `coolingrate_proper.c`. @ChunYen-Chen Do you agree?

hyschive commented:
@ChunYen-Chen After merging this PR, please help file a follow-up PR to address #354. Thanks.

YuriOku commented on example/test_problem/Hydro/Grackle_Comoving/README:
```diff
@@ -0,0 +1,20 @@
+Compilation flags:
+========================================
+Enable : MODEL=HYDRO, COMOVING, SUPPORT_GRACKLE
+Disable:
```
added `PARTICLE`

YuriOku commented on example/test_problem/Hydro/Grackle_Comoving/clean.sh:
```diff
@@ -0,0 +1,6 @@
+      Diag* BaseXYslice* BaseYZslice* BaseXZslice* BaseXline* BaseYline* BaseZline* BaseDiag* \
+      PowerSpec_* Particle_* nohup.out Record__Performance Record__TimingMPI_* \
+      Record__ParticleCount Record__User Patch_* Record__NCorrUnphy FailedPatchGroup* *.pyc Record__LoadBalance \
+      GRACKLE_INFO Record__DivB Record__Center plot/*png
```
removed `plot/*png`

YuriOku commented on example/test_problem/Hydro/Grackle_Comoving/plot/Compute_CoolingCurve.py:
```diff
@@ -0,0 +1,60 @@
+# compute cooling curve from the time evolution of the temperature
```
added the shebang

YuriOku commented on example/test_problem/Hydro/Grackle_Comoving/Input__DumpTable:
removed the dump table

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+#  if   !( MODEL == HYDRO )
+   Aux_Error( ERROR_INFO, "MODEL != HYDRO !!\n" );
+#  endif
+
```
added the check

YuriOku commented on src/Grackle/Grackle_AdvanceDt.cpp:
```diff
@@ -26,8 +26,16 @@
-
-   InvokeSolver( GRACKLE_SOLVER, lv, TimeNew, TimeOld, dt, NULL_REAL, SaveSg, NULL_INT, NULL_INT, OverlapMPI, Overlap_Sync );
+   double dt_grackle;
+#ifdef COMOVING
+   // convert dt from the comoving time interval to the physical time interval
+   // --> see Equation (15) of Schive, Tsai, & Chiueh (2010)
+   dt_grackle = dt * SQR(TimeOld);
+#else
+   dt_grackle = dt;
+#endif
```
fixed the style

YuriOku commented on example/test_problem/Hydro/Grackle_Comoving/Input__Parameter:
```diff
@@ -0,0 +1,400 @@
+NX0_TOT_Y                     16          # number of base-level cells along y
+NX0_TOT_Z                     16          # number of base-level cells along z
+OMP_NTHREAD                  -1           # number of OpenMP threads (<=0=auto) [-1] ##OPENMP ONLY##
+END_T                         1.0         # end physical time (<0=auto -> must be set by test problems or restart) [-1.0]
```
set `END_T` to `-1.0`

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+   fluid[Idx_Metal] = GrackleComoving_InitialMetallicity * Dens;
+   }
+} // FUNCTION : SetGridIC
+
```
added empty lines to make three empty lines after `SetGridIC`

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+   Che_Units.a_units              = 1.0;
+   Che_Units.a_value              = Time[0];
+
+   // Calculate cooling time.
+   if (calculate_cooling_time(&Che_Units, &my_fields, my_cooling_time) == 0) {
+     Aux_Error( ERROR_INFO, "Error in calculate_cooling_time.\n");
+   }
+   // calculate temperature
+   if (calculate_temperature(&Che_Units, &my_fields, my_temperature) == 0) {
+      Aux_Error( ERROR_INFO, "Error in calculate_temperature.\n");
+   }
+   // calculate gamma
+   if (calculate_gamma(&Che_Units, &my_fields, my_gamma) == 0) {
+      Aux_Error( ERROR_INFO, "Error in calculate_gamma.\n");
+   }
```
updated

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+   SetParameter();
+
+// set the function pointers of various problem-specific routines
+   Init_Function_User_Ptr = SetGridIC;
+   Aux_Record_User_Ptr    = Aux_Record_GrackleComoving;
+   End_User_Ptr           = End_GrackleComoving;
+   Mis_GetTimeStep_User_Ptr = Mis_GetTimeStep_GrackleComoving;
```
aligned `=`

YuriOku commented on include/GAMER.h:
```diff
@@ -57,6 +57,7 @@
 #else
 #  define CONFIG_BFLOAT_4
 #endif
+#define OMIT_LEGACY_INTERNAL_GRACKLE_FUNC
```
added an empty line

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+      Aux_Message( stdout, "=============================================================================\n" );
+   }
+
+   grackle_data->max_iterations = 100000;
```
I have removed the line `grackle_data->max_iterations = 100000;`. This was necessary when the timestep size was too large compared to the cooling time. The timestep size is now limited by the cooling time as computed in `Mis_GetTimeStep_GrackleComoving`, so the line is unnecessary.

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+
+
+// (4) make a note
+   if ( MPI_Rank == 0 )
+   {
+      Aux_Message( stdout, "=============================================================================\n" );
+      Aux_Message( stdout, "  test problem ID = %d\n", TESTPROB_ID  );
+      Aux_Message( stdout, "  InitialTemperature   = %13.7e K\n",         GrackleComoving_InitialTemperature );
+      Aux_Message( stdout, "  InitialMetallicity   = %13.7e\n",         GrackleComoving_InitialMetallicity );
+      Aux_Message( stdout, "=============================================================================\n" );
+   }
```
updated

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+// ********************************************************************************************************************************
+// ReadPara->Add( "KEY_IN_THE_FILE",   &VARIABLE,              DEFAULT,       MIN,              MAX               );
+// ********************************************************************************************************************************
+   ReadPara->Add( "GrackleComoving_InitialTemperature", &GrackleComoving_InitialTemperature,  -1.0,  Eps_double,  NoMax_double );
+   ReadPara->Add( "GrackleComoving_InitialMetallicity", &GrackleComoving_InitialMetallicity,  0.0,   0.0,  NoMax_double );
```
updated

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+      double Temp = my_temperature[0];
+      double Lcool = Edens / fabs(my_cooling_time[0] * UNIT_T) / n / n;
+
+      fprintf( File_User, "%14.7e%14ld%3s%14.7e%14.7e%14.7e%14.7e%14.7e%14.7e", Time[0], Step, "", dt_SubStep, n, mu, Temp, Edens, Lcool);
```
added a space before `)`

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+// user-specified info
+   if ( MPI_Rank == 0 )
+   {
+      FILE *File_User = fopen( FileName, "a" );
+      int FluSg = amr->FluSg[0];
+      double Dens = amr->patch[FluSg][0][0]->fluid[DENS][0][0][0];
+      double Eint = amr->patch[FluSg][0][0]->fluid[ENGY][0][0][0]; // assume no magnetic and kinetic energy
```
aligned

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+//-------------------------------------------------------------------------------------------------------
+void End_GrackleComoving()
+{
+   delete [] my_fields.grid_dimension;
+   delete [] my_fields.grid_start;
+   delete [] my_fields.grid_end;
+
+   my_fields.grid_dimension = NULL;
+   my_fields.grid_start = NULL;
+   my_fields.grid_end = NULL;
```
followed your suggestion and updated the code

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+      Aux_Message( stdout, "=============================================================================\n" );
+   }
+
+   grackle_data->max_iterations = 100000;
```
This line has been removed

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+   // set temperature units
+   double temperature_units = get_temperature_units(&Che_Units);
+   double u = GrackleComoving_InitialTemperature / (mu * (GAMMA - 1.) * temperature_units) * SQR(Time);
+   Dens = 1.0;
+   MomX = 0.0;
+   MomY = 0.0;
+   MomZ = 0.0;
+   Eint = Dens * u;
+   Etot = Hydro_ConEint2Etot( Dens, MomX, MomY, MomZ, Eint, 0.0 );     // do NOT include magnetic energy here
+
+   fluid[DENS] = Dens;
+   fluid[MOMX] = MomX;
+   fluid[MOMY] = MomY;
+   fluid[MOMZ] = MomZ;
+   fluid[ENGY] = Etot;
+   
+// initialize all chemical species
+   if ( GRACKLE_PRIMORDIAL >= GRACKLE_PRI_CHE_NSPE6 ) {
+   fluid[Idx_HI   ] = 0.0;
+   fluid[Idx_HII  ] = grackle_data->HydrogenFractionByMass * (1.0 - GrackleComoving_InitialMetallicity) * Dens;
+   fluid[Idx_HeI  ] = 0.0;
+   fluid[Idx_HeII ] = 0.0;
+   fluid[Idx_HeIII] = (1.0 - grackle_data->HydrogenFractionByMass) * (1.0 - GrackleComoving_InitialMetallicity) * Dens;
+   fluid[Idx_e    ] = (fluid[Idx_HII  ] + fluid[Idx_HeII ] / 4.0 + 2.0 * fluid[Idx_HeIII] / 4.0) * Const_me / Const_mp;
+   }
+// 9-species network
+   if ( GRACKLE_PRIMORDIAL >= GRACKLE_PRI_CHE_NSPE9 ) {
+   fluid[Idx_HM   ] = 0.0;
+   fluid[Idx_H2I  ] = 0.0;
+   fluid[Idx_H2II ] = 0.0;
+   }
+// 12-species network
+   if ( GRACKLE_PRIMORDIAL >= GRACKLE_PRI_CHE_NSPE12 ) {
+   fluid[Idx_DI   ] = 0.0;
+   fluid[Idx_DII  ] = grackle_data->DeuteriumToHydrogenRatio * fluid[Idx_HI];
+   fluid[Idx_HDI  ] = 0.0;
+   }
+   if ( GRACKLE_METAL ) {
+   fluid[Idx_Metal] = GrackleComoving_InitialMetallicity * Dens;
+   }
```
wrapped those lines with the preprocessor directives

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+   // Set grid dimension and size.
+   // grid_start and grid_end are used to ignore ghost zones.
+   int field_size = 1;
+   my_fields.grid_rank = 3;
+   my_fields.grid_dimension = new int[3];
+   my_fields.grid_start = new int[3];
+   my_fields.grid_end = new int[3];
+   for (int i = 0;i < 3;i++) {
+     my_fields.grid_dimension[i] = 1; // the active dimension not including ghost zones.
+     my_fields.grid_start[i] = 0;
+     my_fields.grid_end[i] = 0;
+   }
+   my_fields.grid_dimension[0] = field_size;
+   my_fields.grid_end[0] = field_size - 1;
+   my_fields.grid_dx = 0.0; // used only for H2 self-shielding approximation
+
+   my_fields.density         = new gr_float[field_size];
+   my_fields.internal_energy = new gr_float[field_size];
+   // for primordial_chemistry >= 1
+   my_fields.HI_density      = new gr_float[field_size];
+   my_fields.HII_density     = new gr_float[field_size];
+   my_fields.HeI_density     = new gr_float[field_size];
+   my_fields.HeII_density    = new gr_float[field_size];
+   my_fields.HeIII_density   = new gr_float[field_size];
+   my_fields.e_density       = new gr_float[field_size];
+   // for primordial_chemistry >= 2
+   my_fields.HM_density      = new gr_float[field_size];
+   my_fields.H2I_density     = new gr_float[field_size];
+   my_fields.H2II_density    = new gr_float[field_size];
+   // for primordial_chemistry >= 3
+   my_fields.DI_density      = new gr_float[field_size];
+   my_fields.DII_density     = new gr_float[field_size];
+   my_fields.HDI_density     = new gr_float[field_size];
+   // for metal_cooling = 1
+   my_fields.metal_density   = new gr_float[field_size];
+
+   my_temperature  = new gr_float[field_size];;
+   my_gamma        = new gr_float[field_size];;
+   my_cooling_time = new gr_float[field_size];;
```
wrapped the lines with the preprocessor directives

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+   // determine the time-step
+   double dTime_user = HUGE_NUMBER;
+
+   // Calculate cooling time.
+   if (calculate_cooling_time(&Che_Units, &my_fields, my_cooling_time) == 0) {
+     Aux_Error( ERROR_INFO, "Error in calculate_cooling_time.\n");
+   }
+   
+   dTime_user = FMIN(dTime_user, 0.01 * fabs(my_cooling_time[0]));
+
+   my_fields.internal_energy[0] *= 0.9;
+
+   // Recalculate cooling time.
+   if (calculate_cooling_time(&Che_Units, &my_fields, my_cooling_time) == 0) {
+     Aux_Error( ERROR_INFO, "Error in calculate_cooling_time.\n");
+   }
+   
+   dTime_user = FMIN(dTime_user, 0.01 * fabs(my_cooling_time[0]));
```
wrapped the lines with preprocessor directives

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+
+   double MassRatio_pe    = Const_mp / Const_me;
+
+   my_fields.density[0] = Dens;
+   my_fields.internal_energy[0] = Eint / Dens / SQR(Time[lv]);
+
+   if ( GRACKLE_PRIMORDIAL >= GRACKLE_PRI_CHE_NSPE6 ) {
+      my_fields.e_density    [0] = amr->patch[FluSg][0][0]->fluid[Idx_e    ][0][0][0] * MassRatio_pe;
+      my_fields.HI_density   [0] = amr->patch[FluSg][0][0]->fluid[Idx_HI   ][0][0][0] ;
+      my_fields.HII_density  [0] = amr->patch[FluSg][0][0]->fluid[Idx_HII  ][0][0][0] ;
+      my_fields.HeI_density  [0] = amr->patch[FluSg][0][0]->fluid[Idx_HeI  ][0][0][0] ;
+      my_fields.HeII_density [0] = amr->patch[FluSg][0][0]->fluid[Idx_HeII ][0][0][0] ;
+      my_fields.HeIII_density[0] = amr->patch[FluSg][0][0]->fluid[Idx_HeIII][0][0][0] ;
+   }
+
+   if ( GRACKLE_PRIMORDIAL >= GRACKLE_PRI_CHE_NSPE9 ) {
+      my_fields.HM_density  [0]  = amr->patch[FluSg][0][0]->fluid[Idx_HM  ][0][0][0] ;
+      my_fields.H2I_density [0]  = amr->patch[FluSg][0][0]->fluid[Idx_H2I ][0][0][0] ;
+      my_fields.H2II_density[0]  = amr->patch[FluSg][0][0]->fluid[Idx_H2II][0][0][0] ;
+   }
+
+   if ( GRACKLE_PRIMORDIAL >= GRACKLE_PRI_CHE_NSPE12 ) {
+      my_fields.DI_density [0]   = amr->patch[FluSg][0][0]->fluid[Idx_DI  ][0][0][0] ;
+      my_fields.DII_density[0]   = amr->patch[FluSg][0][0]->fluid[Idx_DII ][0][0][0] ;
+      my_fields.HDI_density[0]   = amr->patch[FluSg][0][0]->fluid[Idx_HDI ][0][0][0] ;
+   }
+
+   if ( GRACKLE_METAL ) {
+      my_fields.metal_density[0] = amr->patch[FluSg][0][0]->fluid[Idx_Metal][0][0][0] ;
+   }
+
+   Che_Units.comoving_coordinates = 1;
+   Che_Units.density_units        = UNIT_D / CUBE(Time[lv]);
+   Che_Units.length_units         = UNIT_L * Time[lv];
+   Che_Units.time_units           = UNIT_T;
+   Che_Units.velocity_units       = UNIT_V;
+   Che_Units.a_units              = 1.0;
+   Che_Units.a_value              = Time[lv];
```
wrapped the lines with the preprocessor directives

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+      my_fields.DII_density[0]   = amr->patch[FluSg][0][0]->fluid[Idx_DII ][0][0][0] ;
+      my_fields.HDI_density[0]   = amr->patch[FluSg][0][0]->fluid[Idx_HDI ][0][0][0] ;
+   }
+
+   if ( GRACKLE_METAL ) {
+      my_fields.metal_density[0] = amr->patch[FluSg][0][0]->fluid[Idx_Metal][0][0][0] ;
+   }
+
+   Che_Units.comoving_coordinates = 1;
+   Che_Units.density_units        = UNIT_D / CUBE(Time[0]);
+   Che_Units.length_units         = UNIT_L * Time[0];
+   Che_Units.time_units           = UNIT_T;
+   Che_Units.velocity_units       = UNIT_V;
+   Che_Units.a_units              = 1.0;
+   Che_Units.a_value              = Time[0];
+
+   // Calculate cooling time.
+   if (calculate_cooling_time(&Che_Units, &my_fields, my_cooling_time) == 0) {
+     Aux_Error( ERROR_INFO, "Error in calculate_cooling_time.\n");
+   }
+   // calculate temperature
+   if (calculate_temperature(&Che_Units, &my_fields, my_temperature) == 0) {
+      Aux_Error( ERROR_INFO, "Error in calculate_temperature.\n");
+   }
+   // calculate gamma
+   if (calculate_gamma(&Che_Units, &my_fields, my_gamma) == 0) {
+      Aux_Error( ERROR_INFO, "Error in calculate_gamma.\n");
+   }
+
+      double dt_SubStep = Mis_dTime2dt( Time[0], dTime_Base ) * SQR(Time[0]) * UNIT_T;
+
+      // set temperature units
+      double temperature_units = get_temperature_units(&Che_Units);
+      double mu = my_temperature[0] / (my_fields.internal_energy[0] * (my_gamma[0] - 1.) * temperature_units);
+      double n = Dens / CUBE(Time[0]) * UNIT_D / mu / Const_mp;
+      double Edens = Eint * UNIT_P / SQR(Time[0]) / CUBE(Time[0]);
+      double Temp = my_temperature[0];
+      double Lcool = Edens / fabs(my_cooling_time[0] * UNIT_T) / n / n;
+
+      fprintf( File_User, "%14.7e%14ld%3s%14.7e%14.7e%14.7e%14.7e%14.7e%14.7e", Time[0], Step, "", dt_SubStep, n, mu, Temp, Edens, Lcool);
```
wrapped the lines with preprocessor directives

YuriOku commented on src/Init/Init_Field.cpp:
```diff
@@ -76,8 +76,10 @@ void Init_Field()
+// --> see the subroutine make_consistent_g in solve_rate_cool_g.F in the Grackle source code
    if ( GRACKLE_PRIMORDIAL >= GRACKLE_PRI_CHE_NSPE6 ) {
-   Idx_e       = AddField( "Electron", FIXUP_FLUX_YES, FIXUP_REST_YES, NORMALIZE_YES, INTERP_FRAC_YES );
+   Idx_e       = AddField( "Electron", FIXUP_FLUX_YES, FIXUP_REST_YES, NORMALIZE_NO , INTERP_FRAC_YES ); // electron mass is neglected in Grackle
```
fixed

YuriOku commented on src/Init/Init_Field.cpp:
```diff
@@ -92,9 +94,9 @@ void Init_Field()
-   Idx_DI      = AddField( "DI",       FIXUP_FLUX_YES, FIXUP_REST_YES, NORMALIZE_YES, INTERP_FRAC_YES );
-   Idx_DII     = AddField( "DII",      FIXUP_FLUX_YES, FIXUP_REST_YES, NORMALIZE_YES, INTERP_FRAC_YES );
-   Idx_HDI     = AddField( "HDI",      FIXUP_FLUX_YES, FIXUP_REST_YES, NORMALIZE_YES, INTERP_FRAC_YES );
+   Idx_DI      = AddField( "DI",       FIXUP_FLUX_YES, FIXUP_REST_YES, NORMALIZE_NO , INTERP_FRAC_YES ); // deuterium mass is neglected in Grackle
+   Idx_DII     = AddField( "DII",      FIXUP_FLUX_YES, FIXUP_REST_YES, NORMALIZE_NO , INTERP_FRAC_YES ); // deuterium mass is neglected in Grackle
+   Idx_HDI     = AddField( "HDI",      FIXUP_FLUX_YES, FIXUP_REST_YES, NORMALIZE_NO , INTERP_FRAC_YES ); // deuterium mass is neglected in Grackle
```
fixed

YuriOku commented on src/Grackle/Grackle_AdvanceDt.cpp:
```diff
@@ -26,8 +26,16 @@
 {
-
-   InvokeSolver( GRACKLE_SOLVER, lv, TimeNew, TimeOld, dt, NULL_REAL, SaveSg, NULL_INT, NULL_INT, OverlapMPI, Overlap_Sync );
+   double dt_grackle;
+#ifdef COMOVING
+   // convert dt from the comoving time interval to the physical time interval
+   // --> see Equation (15) of Schive, Tsai, & Chiueh (2010)
+   dt_grackle = dt * SQR(TimeOld);
+#else
+   dt_grackle = dt;
+#endif
```
followed your comment and defined `dt_grackle` as a constant

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+// =======================================================================================
+static double GrackleComoving_InitialTemperature;
+static double GrackleComoving_InitialMetallicity;
+grackle_field_data my_fields;
```
followed your comment and declared `my_fields` as a `static`

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+
+// problem-specific global variables
+// =======================================================================================
+static double GrackleComoving_InitialTemperature;
+static double GrackleComoving_InitialMetallicity;
+grackle_field_data my_fields;
+static gr_float *my_temperature;
+static gr_float *my_gamma;
+static gr_float *my_cooling_time;
```
aligned

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+//-------------------------------------------------------------------------------------------------------
+double Mis_GetTimeStep_GrackleComoving( const int lv, const double dTime_dt )
+{
+   int FluSg = amr->FluSg[0];
+   double Dens = amr->patch[FluSg][0][0]->fluid[DENS][0][0][0];
+   double Eint = amr->patch[FluSg][0][0]->fluid[ENGY][0][0][0]; // assume no magnetic and kinetic energy
```
aligned

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+
+#  endif // #ifdef DUAL_ENERGY
+
+   double MassRatio_pe    = Const_mp / Const_me;
```
declared `MassRatio_pe` as a `const`

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+   my_fields.density[0] = Dens;
+   my_fields.internal_energy[0] = Eint / Dens / SQR(Time[lv]);
+
+   if ( GRACKLE_PRIMORDIAL >= GRACKLE_PRI_CHE_NSPE6 ) {
+      my_fields.e_density    [0] = amr->patch[FluSg][0][0]->fluid[Idx_e    ][0][0][0] * MassRatio_pe;
+      my_fields.HI_density   [0] = amr->patch[FluSg][0][0]->fluid[Idx_HI   ][0][0][0] ;
+      my_fields.HII_density  [0] = amr->patch[FluSg][0][0]->fluid[Idx_HII  ][0][0][0] ;
+      my_fields.HeI_density  [0] = amr->patch[FluSg][0][0]->fluid[Idx_HeI  ][0][0][0] ;
+      my_fields.HeII_density [0] = amr->patch[FluSg][0][0]->fluid[Idx_HeII ][0][0][0] ;
+      my_fields.HeIII_density[0] = amr->patch[FluSg][0][0]->fluid[Idx_HeIII][0][0][0] ;
+   }
+
+   if ( GRACKLE_PRIMORDIAL >= GRACKLE_PRI_CHE_NSPE9 ) {
+      my_fields.HM_density  [0]  = amr->patch[FluSg][0][0]->fluid[Idx_HM  ][0][0][0] ;
+      my_fields.H2I_density [0]  = amr->patch[FluSg][0][0]->fluid[Idx_H2I ][0][0][0] ;
+      my_fields.H2II_density[0]  = amr->patch[FluSg][0][0]->fluid[Idx_H2II][0][0][0] ;
+   }
+
+   if ( GRACKLE_PRIMORDIAL >= GRACKLE_PRI_CHE_NSPE12 ) {
+      my_fields.DI_density [0]   = amr->patch[FluSg][0][0]->fluid[Idx_DI  ][0][0][0] ;
+      my_fields.DII_density[0]   = amr->patch[FluSg][0][0]->fluid[Idx_DII ][0][0][0] ;
+      my_fields.HDI_density[0]   = amr->patch[FluSg][0][0]->fluid[Idx_HDI ][0][0][0] ;
+   }
+
+   if ( GRACKLE_METAL ) {
+      my_fields.metal_density[0] = amr->patch[FluSg][0][0]->fluid[Idx_Metal][0][0][0] ;
+   }
```
updated

YuriOku commented on src/Grackle/Grackle_Init.cpp:
```diff
@@ -52,11 +47,16 @@ void Grackle_Init()
 // units in cgs
 // --> Che_Units is declared as a global variable since all Grackle solvers require that as well
 #  ifdef COMOVING
+   // see https://grackle.readthedocs.io/en/latest/Interaction.html#comoving-coordinates and
+   // https://github.com/grackle-project/grackle/issues/192
+   // --> the density and length units are conversion factors from code values to cgs in proper coordinates,
+   //     while the time and velocity units should remain constant, i.e., velocity_units = length_units / (Time * time_units)
+   // --> the specific energy passed to Grackle is u = k_B * T / ((gamma - 1) mu m_p)
```
updated

YuriOku commented on src/Grackle/Grackle_Init.cpp:
```diff
@@ -52,11 +47,16 @@ void Grackle_Init()
+   // https://github.com/grackle-project/grackle/issues/192
+   // --> the density and length units are conversion factors from code values to cgs in proper coordinates,
+   //     while the time and velocity units should remain constant, i.e., velocity_units = length_units / (Time * time_units)
+   // --> the specific energy passed to Grackle is u = k_B * T / ((gamma - 1) mu m_p)
```
fixed typo

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+   // grid_start and grid_end are used to ignore ghost zones.
+   int field_size = 1;
+   my_fields.grid_rank = 3;
+   my_fields.grid_dimension = new int[3];
+   my_fields.grid_start = new int[3];
+   my_fields.grid_end = new int[3];
```
aligned `=`

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+   my_fields.grid_start = new int[3];
+   my_fields.grid_end = new int[3];
+   for (int i = 0;i < 3;i++) {
+     my_fields.grid_dimension[i] = 1; // the active dimension not including ghost zones.
+     my_fields.grid_start[i] = 0;
+     my_fields.grid_end[i] = 0;
```
followed the style

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+   my_fields.grid_end[0] = field_size - 1;
+   my_fields.grid_dx = 0.0; // used only for H2 self-shielding approximation
+
+   my_fields.density         = new gr_float[field_size];
+   my_fields.internal_energy = new gr_float[field_size];
+   // for primordial_chemistry >= 1
+   my_fields.HI_density      = new gr_float[field_size];
+   my_fields.HII_density     = new gr_float[field_size];
+   my_fields.HeI_density     = new gr_float[field_size];
+   my_fields.HeII_density    = new gr_float[field_size];
+   my_fields.HeIII_density   = new gr_float[field_size];
+   my_fields.e_density       = new gr_float[field_size];
+   // for primordial_chemistry >= 2
+   my_fields.HM_density      = new gr_float[field_size];
+   my_fields.H2I_density     = new gr_float[field_size];
+   my_fields.H2II_density    = new gr_float[field_size];
+   // for primordial_chemistry >= 3
+   my_fields.DI_density      = new gr_float[field_size];
+   my_fields.DII_density     = new gr_float[field_size];
+   my_fields.HDI_density     = new gr_float[field_size];
+   // for metal_cooling = 1
+   my_fields.metal_density   = new gr_float[field_size];
+
+   my_temperature  = new gr_float[field_size];;
+   my_gamma        = new gr_float[field_size];;
+   my_cooling_time = new gr_float[field_size];;
```
updated

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+   my_fields.grid_dimension = new int[3];
+   my_fields.grid_start = new int[3];
+   my_fields.grid_end = new int[3];
+   for (int i = 0;i < 3;i++) {
```
updated

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+
+   // Set grid dimension and size.
+   // grid_start and grid_end are used to ignore ghost zones.
+   int field_size = 1;
```
declared `field_size` as a `const`

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+   double Dual = amr->patch[FluSg][0][0]->fluid[DUAL][0][0][0];
+
+#  if   ( DUAL_ENERGY == DE_ENPY )
+   Pres  = Hydro_DensDual2Pres( Dens, Dual, EoS_AuxArray_Flt[1], CheckMinPres_No, NULL_REAL );
```
added `double` before `Pres` and declared `const bool CheckMinPres_No = false`

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+      fprintf( File_User, "\n" );
+
+      fclose( File_User );
+   }
```
added the comment

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+      my_fields.HeII_density [0] = amr->patch[FluSg][0][0]->fluid[Idx_HeII ][0][0][0] ;
+      my_fields.HeIII_density[0] = amr->patch[FluSg][0][0]->fluid[Idx_HeIII][0][0][0] ;
+   }
+
+   if ( GRACKLE_PRIMORDIAL >= GRACKLE_PRI_CHE_NSPE9 ) {
+      my_fields.HM_density  [0]  = amr->patch[FluSg][0][0]->fluid[Idx_HM  ][0][0][0] ;
+      my_fields.H2I_density [0]  = amr->patch[FluSg][0][0]->fluid[Idx_H2I ][0][0][0] ;
+      my_fields.H2II_density[0]  = amr->patch[FluSg][0][0]->fluid[Idx_H2II][0][0][0] ;
+   }
+
+   if ( GRACKLE_PRIMORDIAL >= GRACKLE_PRI_CHE_NSPE12 ) {
+      my_fields.DI_density [0]   = amr->patch[FluSg][0][0]->fluid[Idx_DI  ][0][0][0] ;
+      my_fields.DII_density[0]   = amr->patch[FluSg][0][0]->fluid[Idx_DII ][0][0][0] ;
+      my_fields.HDI_density[0]   = amr->patch[FluSg][0][0]->fluid[Idx_HDI ][0][0][0] ;
+   }
+
+   if ( GRACKLE_METAL ) {
+      my_fields.metal_density[0] = amr->patch[FluSg][0][0]->fluid[Idx_Metal][0][0][0] ;
+   }
+
+   Che_Units.comoving_coordinates = 1;
+   Che_Units.density_units        = UNIT_D / CUBE(Time[0]);
+   Che_Units.length_units         = UNIT_L * Time[0];
+   Che_Units.time_units           = UNIT_T;
+   Che_Units.velocity_units       = UNIT_V;
+   Che_Units.a_units              = 1.0;
+   Che_Units.a_value              = Time[0];
+
+   // Calculate cooling time.
+   if (calculate_cooling_time(&Che_Units, &my_fields, my_cooling_time) == 0) {
+     Aux_Error( ERROR_INFO, "Error in calculate_cooling_time.\n");
+   }
+   // calculate temperature
+   if (calculate_temperature(&Che_Units, &my_fields, my_temperature) == 0) {
+      Aux_Error( ERROR_INFO, "Error in calculate_temperature.\n");
+   }
+   // calculate gamma
+   if (calculate_gamma(&Che_Units, &my_fields, my_gamma) == 0) {
+      Aux_Error( ERROR_INFO, "Error in calculate_gamma.\n");
+   }
```
fixed the indent

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+
+   delete [] my_fields.density         ;
+   delete [] my_fields.internal_energy ;
+   // for primordial_chemistry >= 1
+   delete [] my_fields.HI_density      ;
+   delete [] my_fields.HII_density     ;
+   delete [] my_fields.HeI_density     ;
+   delete [] my_fields.HeII_density    ;
+   delete [] my_fields.HeIII_density   ;
+   delete [] my_fields.e_density       ;
+   // for primordial_chemistry >= 2
+   delete [] my_fields.HM_density      ;
+   delete [] my_fields.H2I_density     ;
+   delete [] my_fields.H2II_density    ;
+   // for primordial_chemistry >= 3
+   delete [] my_fields.DI_density      ;
+   delete [] my_fields.DII_density     ;
+   delete [] my_fields.HDI_density     ;
+   // for metal_cooling = 1
+   delete [] my_fields.metal_density   ;
+
+   my_fields.density         = NULL;
+   my_fields.internal_energy = NULL;
+   // for primordial_chemistry >= 1
+   my_fields.HI_density      = NULL;
+   my_fields.HII_density     = NULL;
+   my_fields.HeI_density     = NULL;
+   my_fields.HeII_density    = NULL;
+   my_fields.HeIII_density   = NULL;
+   my_fields.e_density       = NULL;
+   // for primordial_chemistry >= 2
+   my_fields.HM_density      = NULL;
+   my_fields.H2I_density     = NULL;
+   my_fields.H2II_density    = NULL;
+   // for primordial_chemistry >= 3
+   my_fields.DI_density      = NULL;
+   my_fields.DII_density     = NULL;
+   my_fields.HDI_density     = NULL;
+   // for metal_cooling = 1
+   my_fields.metal_density   = NULL;
```
wrapped those lines with the preprocessor directives

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+   my_fields.density[0] = Dens;
+   my_fields.internal_energy[0] = Eint / Dens / SQR(Time[0]);
+
+   if ( GRACKLE_PRIMORDIAL >= GRACKLE_PRI_CHE_NSPE6 ) {
+      my_fields.e_density    [0] = amr->patch[FluSg][0][0]->fluid[Idx_e    ][0][0][0]  * MassRatio_pe;
+      my_fields.HI_density   [0] = amr->patch[FluSg][0][0]->fluid[Idx_HI   ][0][0][0] ;
+      my_fields.HII_density  [0] = amr->patch[FluSg][0][0]->fluid[Idx_HII  ][0][0][0] ;
+      my_fields.HeI_density  [0] = amr->patch[FluSg][0][0]->fluid[Idx_HeI  ][0][0][0] ;
+      my_fields.HeII_density [0] = amr->patch[FluSg][0][0]->fluid[Idx_HeII ][0][0][0] ;
+      my_fields.HeIII_density[0] = amr->patch[FluSg][0][0]->fluid[Idx_HeIII][0][0][0] ;
+   }
+
+   if ( GRACKLE_PRIMORDIAL >= GRACKLE_PRI_CHE_NSPE9 ) {
+      my_fields.HM_density  [0]  = amr->patch[FluSg][0][0]->fluid[Idx_HM  ][0][0][0] ;
+      my_fields.H2I_density [0]  = amr->patch[FluSg][0][0]->fluid[Idx_H2I ][0][0][0] ;
+      my_fields.H2II_density[0]  = amr->patch[FluSg][0][0]->fluid[Idx_H2II][0][0][0] ;
+   }
+
+   if ( GRACKLE_PRIMORDIAL >= GRACKLE_PRI_CHE_NSPE12 ) {
+      my_fields.DI_density [0]   = amr->patch[FluSg][0][0]->fluid[Idx_DI  ][0][0][0] ;
+      my_fields.DII_density[0]   = amr->patch[FluSg][0][0]->fluid[Idx_DII ][0][0][0] ;
+      my_fields.HDI_density[0]   = amr->patch[FluSg][0][0]->fluid[Idx_HDI ][0][0][0] ;
+   }
+
+   if ( GRACKLE_METAL ) {
+      my_fields.metal_density[0] = amr->patch[FluSg][0][0]->fluid[Idx_Metal][0][0][0] ;
+   }
```
updated

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+      Aux_Error( ERROR_INFO, "Error in calculate_gamma.\n");
+   }
+
+      double dt_SubStep = Mis_dTime2dt( Time[0], dTime_Base ) * SQR(Time[0]) * UNIT_T;
+
+      // set temperature units
+      double temperature_units = get_temperature_units(&Che_Units);
+      double mu = my_temperature[0] / (my_fields.internal_energy[0] * (my_gamma[0] - 1.) * temperature_units);
+      double n = Dens / CUBE(Time[0]) * UNIT_D / mu / Const_mp;
+      double Edens = Eint * UNIT_P / SQR(Time[0]) / CUBE(Time[0]);
+      double Temp = my_temperature[0];
+      double Lcool = Edens / fabs(my_cooling_time[0] * UNIT_T) / n / n;
```
declared them as const

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+      Aux_Error( ERROR_INFO, "Error in calculate_gamma.\n");
+   }
+
+      double dt_SubStep = Mis_dTime2dt( Time[0], dTime_Base ) * SQR(Time[0]) * UNIT_T;
+
+      // set temperature units
+      double temperature_units = get_temperature_units(&Che_Units);
+      double mu = my_temperature[0] / (my_fields.internal_energy[0] * (my_gamma[0] - 1.) * temperature_units);
+      double n = Dens / CUBE(Time[0]) * UNIT_D / mu / Const_mp;
+      double Edens = Eint * UNIT_P / SQR(Time[0]) / CUBE(Time[0]);
+      double Temp = my_temperature[0];
+      double Lcool = Edens / fabs(my_cooling_time[0] * UNIT_T) / n / n;
```
aligned

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+{
+   double Dens, MomX, MomY, MomZ, Eint, Etot;
+   
+   double mu = 4 / (8 - 5 * (1 - grackle_data->HydrogenFractionByMass)); // fully ionized gas
+   // set temperature units
+   double temperature_units = get_temperature_units(&Che_Units);
+   double u = GrackleComoving_InitialTemperature / (mu * (GAMMA - 1.) * temperature_units) * SQR(Time);
```
aligned

YuriOku commented on example/test_problem/Hydro/Grackle_Comoving/plot/coolingrate_proper.c:
Same with me and I think this is a bug in Grackle.

YuriOku replied:
I have added a description in the README
`6. It is recommended to use double-float precision for the Grackle library`

YuriOku commented on example/test_problem/Hydro/Grackle_Comoving/Input__Parameter:
```diff
@@ -0,0 +1,400 @@
+PAR_TR_VEL_CORR               0           # correct tracer particle velocities in regions of discontinuous flow [0]
+
+# cosmology (COMOVING only)
+A_INIT                        0.1        # initial scale factor
```
aligned

YuriOku commented on example/test_problem/Hydro/Grackle_Comoving/Input__Parameter:
```diff
@@ -0,0 +1,400 @@
+
+
+# data dump
+OPT__OUTPUT_TOTAL             1           # output the simulation snapshot: (0=off, 1=HDF5, 2=C-binary) [1]
```
followed your suggestion and disabled the output

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+
+   delete [] my_fields.density         ;
+   delete [] my_fields.internal_energy ;
+   // for primordial_chemistry >= 1
```
removed the comment. this comment was not meaningful.

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+   delete [] my_fields.HeII_density    ;
+   delete [] my_fields.HeIII_density   ;
+   delete [] my_fields.e_density       ;
+   // for primordial_chemistry >= 2
```
removed the comment

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+   delete [] my_fields.HM_density      ;
+   delete [] my_fields.H2I_density     ;
+   delete [] my_fields.H2II_density    ;
+   // for primordial_chemistry >= 3
```
removed the comment

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+   delete [] my_fields.DI_density      ;
+   delete [] my_fields.DII_density     ;
+   delete [] my_fields.HDI_density     ;
+   // for metal_cooling = 1
```
removed the comment

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+   Che_Units.a_units              = 1.0;
+   Che_Units.a_value              = Time[lv];
+
+   // determine the time-step
```
updated

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+   // determine the time-step
+   double dTime_user = HUGE_NUMBER;
+
+   // Calculate cooling time.
+   if (calculate_cooling_time(&Che_Units, &my_fields, my_cooling_time) == 0) {
+     Aux_Error( ERROR_INFO, "Error in calculate_cooling_time.\n");
+   }
```
updated

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+
+   my_fields.internal_energy[0] *= 0.9;
+
+   // Recalculate cooling time.
+   if (calculate_cooling_time(&Che_Units, &my_fields, my_cooling_time) == 0) {
+     Aux_Error( ERROR_INFO, "Error in calculate_cooling_time.\n");
+   }
```
updated

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+// Return      :  dt
+//-------------------------------------------------------------------------------------------------------
+double Mis_GetTimeStep_GrackleComoving( const int lv, const double dTime_dt )
+{
```
added an empty line after the `{`

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+   if ( MPI_Rank == 0 )    Aux_Message( stdout, "%s ...\n", __FUNCTION__ );
+
+
+   // Set grid dimension and size.
+   // grid_start and grid_end are used to ignore ghost zones.
```
updated

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+   fluid[Idx_HeI  ] = 0.0;
+   fluid[Idx_HeII ] = 0.0;
+   fluid[Idx_HeIII] = (1.0 - grackle_data->HydrogenFractionByMass) * (1.0 - GrackleComoving_InitialMetallicity) * Dens;
+   fluid[Idx_e    ] = (fluid[Idx_HII  ] + fluid[Idx_HeII ] / 4.0 + 2.0 * fluid[Idx_HeIII] / 4.0) * Const_me / Const_mp;
```
updated

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+//                2. Invoked by Mis_GetTimeStep() using the function pointer "Mis_GetTimeStep_User_Ptr",
+//                   which must be set by a test problem initializer
+//                3. Enabled by the runtime option "OPT__DT_USER"
+//                4. If COMOVING is off, dTime = (physical time interval) / scale_factor^2
```
removed the note `4.`

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+
+#  endif // #ifdef DUAL_ENERGY
+
+   double MassRatio_pe    = Const_mp / Const_me;
```
declared it as const

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,576 @@
+{
+   double Dens, MomX, MomY, MomZ, Eint, Etot;
+   
+   double mu = 4 / (8 - 5 * (1 - grackle_data->HydrogenFractionByMass)); // fully ionized gas
+   // set temperature units
+   double temperature_units = get_temperature_units(&Che_Units);
+   double u = GrackleComoving_InitialTemperature / (mu * (GAMMA - 1.) * temperature_units) * SQR(Time);
```
declared them as const

YuriOku commented on example/test_problem/Hydro/Grackle_Comoving/plot/Makefile:
```diff
@@ -0,0 +1,25 @@
+all: coolingrate_proper
+
+coolingrate_proper: coolingrate_proper.c
+	$(CC) -I$(GRACKLE_DIR)/include -L$(GRACKLE_DIR)/lib -fopenmp -O2 -o coolingrate_proper coolingrate_proper.c -lgrackle -lm
```
No reason, but the computation is not heavy and high optimization is not necessary.

YuriOku commented on example/test_problem/Hydro/Grackle_Comoving/plot/Makefile:
```diff
@@ -0,0 +1,25 @@
+GRACKLE_DIR = /usr
+CC = gcc
+
+all: coolingrate_proper
```
followed your suggestion and replaced `coolingrate_proper` to `$(EXE)`

YuriOku commented on example/test_problem/Hydro/Grackle_Comoving/README:
```diff
@@ -0,0 +1,20 @@
+   --> limit the timestep by the cooling time
+
+Note:
+========================================
```
added a description
`4. The floating point precision of Grackle must be same with GAMER to compile the code successfully`

YuriOku commented on example/test_problem/Hydro/Grackle_Comoving/README:
```diff
@@ -0,0 +1,20 @@
+========================================
+1. Compute the thermal evolution on comoving coordinate with the Grackle library
+2. The C program plot/coolingrate_proper.c computes the cooling rate in the proper coordinate
+3. The python script plot/Compute_CoolingCurve.py computes the cooling rate from the simulation output and compares it with result of plot/coolingrate_proper.c
```
I have modified the sentence to clarify how to use the scripts.

YuriOku commented:
@ChunYen-Chen 
I removed the trailing space and added a script to download `CloudyData_noUVB.h5`. I also updated the code according to your suggestion, leaving two conversations unresolved that I am not sure I have resolved.

After these fixes, I have confirmed that I can reproduce the same results in the test problem `Grackle_Comoving`.

ChunYen-Chen commented:
@YuriOku Thanks for the update. The PR looks great and I think it is ready to merge.
@hyschive Yes, I do believe the problem is in `coolingrate_proper.c`.

hyschive commented:
@YuriOku Sorry for the delayed progress. We’re planning to merge this PR soon. Could you please update it to resolve the conflicts first?

YuriOku commented:
resolved the conflict

hsinhaoHHuang reviewed (commented):
Hi @YuriOku,
I have finished the review of this PR.
It is exciting to know GAMER can support Grackle with `COMOVING`. Thank you for implementing it.

I have raised some questions and added some suggestions in the comments. Could you help me take a look when you have time?
Thanks!

YuriOku commented on example/test_problem/Hydro/Grackle_Comoving/Input__TestProb:
```diff
@@ -0,0 +1,3 @@
+# problem-specific runtime parameters
+GrackleComoving_InitialTemperature 1e8
+GrackleComoving_InitialMetallicity 1.295e-2   # Zsun
```
provided descriptions for the parameters

YuriOku commented on example/test_problem/Hydro/Grackle_Comoving/generate_make.sh:
```diff
@@ -0,0 +1,6 @@
+PYTHON=python3
+
+${PYTHON} configure.py --machine=eureka_intel --hdf5=true --model=HYDRO --comoving=true --gravity=true --fftw=FFTW3 \
+                       --grackle=true --double=true --passive=1
```
added  "$@"

YuriOku commented on example/test_problem/Template/Input__Parameter:
```diff
@@ -49,6 +49,7 @@ TESTPROB_ID                   0           # test problem ID [0]
                                           #   23: HYDRO MHD Cosmic Ray Diffusion
                                           #  100: HYDRO CDM cosmological simulation (+GRAVITY & COMOVING & PARTICLE)
                                           #  101: HYDRO Zeldovich pancake collapse (+GRAVITY & COMOVING & PARTICLE)
+                                          #  102: HYDRO Grackle comoving (+COMOVING)
```
followed your suggestion

YuriOku commented on include/GAMER.h:
```diff
@@ -58,6 +58,8 @@
 #  define CONFIG_BFLOAT_4
 #endif
 
+#define OMIT_LEGACY_INTERNAL_GRACKLE_FUNC
```
This is to suppress a compile-time warning message of grackle version prior to 3.3.0. 
Please see lines 191-201 of https://github.com/grackle-project/grackle/blob/grackle-3.2.1/src/clib/grackle.h

YuriOku replied:
added the comments for future reference

YuriOku commented on src/Grackle/Grackle_Close.cpp:
```diff
@@ -117,7 +117,12 @@ void Grackle_Close( const int lv, const int SaveSg, const real h_Che_Array[], co
-            Eint = Ptr_sEint[idx_pg]*Dens;
+#           ifdef COMOVING
+//          convert from the proper frame to the comoving specific internal energy
+            Eint = Ptr_sEint[idx_pg] * Dens * SQR(Time[lv]);
```
followed your suggestion for clarity

YuriOku commented on src/Grackle/Grackle_Init.cpp:
```diff
@@ -52,11 +47,16 @@ void Grackle_Init()
+// see https://grackle.readthedocs.io/en/latest/Interaction.html#comoving-coordinates and
+// https://github.com/grackle-project/grackle/issues/192
+// --> the density and length units are conversion factors from code values to cgs in proper coordinates,
+//     while the time and velocity units should remain constant, i.e., velocity_units = length_units / (Time * time_units)
```
followed your suggestion

YuriOku commented on src/Grackle/Grackle_Init.cpp:
```diff
@@ -52,11 +47,16 @@ void Grackle_Init()
+// https://github.com/grackle-project/grackle/issues/192
+// --> the density and length units are conversion factors from code values to cgs in proper coordinates,
+//     while the time and velocity units should remain constant, i.e., velocity_units = length_units / (Time * time_units)
+// --> the specific energy passed to Grackle is u = k_B * T / ((gamma - 1) * mu * m_p)
```
added a reminder that the internal energy passed to grackle is defined on the proper frame.

YuriOku commented on src/Grackle/Grackle_Prepare.cpp:
```diff
@@ -91,6 +92,16 @@ void Grackle_Prepare( const int lv, real h_Che_Array[], const int NPG, const int
+#  endif
 
    const int  Size1pg          = CUBE(PS2);
    const int  Size1v           = NPG*Size1pg;
```
aligned the lines

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,555 @@
+#  endif
+
+
+// warnings
```
removed the line

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,555 @@
+// --> note that VARIABLE, DEFAULT, MIN, and MAX must have the same data type
+// --> some handy constants (e.g., Useless_bool, Eps_double, NoMin_int, ...) are defined in "include/ReadPara.h"
+// ********************************************************************************************************************************
+// ReadPara->Add( "KEY_IN_THE_FILE",   &VARIABLE,              DEFAULT,       MIN,              MAX               );
+// ********************************************************************************************************************************
+   ReadPara->Add( "GrackleComoving_InitialTemperature", &GrackleComoving_InitialTemperature,  -1.0,  Eps_double,   NoMax_double );
+   ReadPara->Add( "GrackleComoving_InitialMetallicity", &GrackleComoving_InitialMetallicity,   0.0,  0.0,          NoMax_double );
```
set the maximum of `GrackleComoving_InitialMetallicity ` to `1.0`

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,555 @@
+// --> note that VARIABLE, DEFAULT, MIN, and MAX must have the same data type
+// --> some handy constants (e.g., Useless_bool, Eps_double, NoMin_int, ...) are defined in "include/ReadPara.h"
+// ********************************************************************************************************************************
+// ReadPara->Add( "KEY_IN_THE_FILE",   &VARIABLE,              DEFAULT,       MIN,              MAX               );
+// ********************************************************************************************************************************
+   ReadPara->Add( "GrackleComoving_InitialTemperature", &GrackleComoving_InitialTemperature,  -1.0,  Eps_double,   NoMax_double );
+   ReadPara->Add( "GrackleComoving_InitialMetallicity", &GrackleComoving_InitialMetallicity,   0.0,  0.0,          NoMax_double );
```
aligned with the header

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,555 @@
+#  ifdef SUPPORT_GRACKLE
+   const double mu                = 4 / (8 - 5 * (1 - grackle_data->HydrogenFractionByMass)); // fully ionized gas
+   const double temperature_units = get_temperature_units(&Che_Units); // set temperature units
+   const double u                 = GrackleComoving_InitialTemperature / (mu * (GAMMA - 1.) * temperature_units) * SQR(Time);
```
added comment about the mean molecular weight

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,555 @@
+void Aux_Record_GrackleComoving()
+{
+
+   const char FileName[] = "Record__User";
```
followed your suggestion

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,555 @@
+         if ( Aux_CheckFileExist(FileName) )    Aux_Message( stderr, "WARNING : file \"%s\" already exists !!\n", FileName );
+
+         FILE *File_User = fopen( FileName, "a" );
+         fprintf( File_User, "#%13s%14s%3s%14s%14s%14s%14s%14s",  "Time", "Step", "", "dt [s]", "n [1/cc]", "mu", "Temp [K]", "Edens [erg/cc], Lcool[erg cm^3 s^-1]" );
```
replaced `Time `by `a_scale`

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,555 @@
+
+
+//-------------------------------------------------------------------------------------------------------
+// Function    :  Mis_GetTimeStep_GrackleComoving
+// Description :  returns 0.1 * cooling time
```
Right. This timestep criterion is to output frequently and draw a cooling function. Resolving cooling time is desirable in general, but it is infeasible.

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,555 @@
+
+   dTime_user = FMIN(dTime_user, 0.01 * fabs(my_cooling_time[0]));
+
+   my_fields.internal_energy[0] *= 0.9;
+
+// recalculate cooling time
```
The grackle function `calculate_cooling_time` returns E/Edot (this is negative if cooling). At T ~ 1e5K to 1e6K, the cooling time gets shorter as the temperature decreases, so the actual cooling time can be shorter than the cooling rate computed by `calculate_cooling_time.` There, the time step size is limited by the cooling time, assuming a slightly lower temperature (internal energy). We could use an implicit scheme to determine the time step size, but the scheme used in the current version would be enough.

YuriOku replied:
added the comment in the code

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,555 @@
+   dTime_user = FMIN(dTime_user, 0.01 * fabs(my_cooling_time[0]));
+#  endif // #ifdef SUPPORT_GRACKLE
+
+   return dTime_user / SQR(Time[lv]);
```
replaced `dTime_user` to `dt_user` and added the comment

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,555 @@
+
+   return dTime_user / SQR(Time[lv]);
+
+} // FUNCTION : Mis_GetTimeStep_GrackleComoving
```
enclosed the content of the function by `#ifdef COMOVING` directive

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,555 @@
+   delete [] my_fields.DI_density;         my_fields.DI_density      = NULL;
+   delete [] my_fields.DII_density;        my_fields.DII_density     = NULL;
+   delete [] my_fields.HDI_density;        my_fields.HDI_density     = NULL;
+   delete [] my_fields.metal_density;      my_fields.metal_density   = NULL;
+#  endif // #ifdef SUPPORT_GRACKLE
```
deleted those fields

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,555 @@
+
+   my_fields.density         = new gr_float [field_size];
+   my_fields.internal_energy = new gr_float [field_size];
+// for primordial_chemistry >= 1
+   my_fields.HI_density      = new gr_float [field_size];
+   my_fields.HII_density     = new gr_float [field_size];
+   my_fields.HeI_density     = new gr_float [field_size];
+   my_fields.HeII_density    = new gr_float [field_size];
+   my_fields.HeIII_density   = new gr_float [field_size];
+   my_fields.e_density       = new gr_float [field_size];
+// for primordial_chemistry >= 2
+   my_fields.HM_density      = new gr_float [field_size];
+   my_fields.H2I_density     = new gr_float [field_size];
+   my_fields.H2II_density    = new gr_float [field_size];
+// for primordial_chemistry >= 3
+   my_fields.DI_density      = new gr_float [field_size];
+   my_fields.DII_density     = new gr_float [field_size];
+   my_fields.HDI_density     = new gr_float [field_size];
+// for metal_cooling = 1
```
added if condition to initialize the field only when necessary

YuriOku commented on example/test_problem/Hydro/Grackle_Comoving/Input__Parameter:
```diff
@@ -0,0 +1,400 @@
+
+
+# test problems
+TESTPROB_ID                  102          # test problem ID [0]
```
aligned

YuriOku commented on example/test_problem/Hydro/Grackle_Comoving/plot/coolingrate_proper.c:
```diff
@@ -0,0 +1,356 @@
+  for (double logT = T_min; logT <= T_max; logT += dT)
+  {
+    for (i = 0;i < field_size;i++) {
+      my_fields.internal_energy[i] = pow(10.0, logT) / temperature_units;
```
This is an initial guess and will be modified in the loop

YuriOku commented on example/test_problem/Hydro/Grackle_Comoving/plot/coolingrate_proper.c:
```diff
@@ -0,0 +1,356 @@
+        return EXIT_FAILURE;
+      }
+
+      double dt = fabs(0.1 * cooling_time[0]);
```
It is because we have only one cell

YuriOku commented on example/test_problem/Hydro/Grackle_Comoving/plot/coolingrate_proper.c:
```diff
@@ -0,0 +1,356 @@
+    my_fields.DI_density[i] = grackle_data->HydrogenFractionByMass * grackle_data->DeuteriumToHydrogenRatio * my_fields.density[i];
+    my_fields.DII_density[i] = tiny_number * my_fields.density[i];
+    my_fields.HDI_density[i] = tiny_number * my_fields.density[i];
+    my_fields.e_density[i] = tiny_number * my_fields.density[i];
```
I have set them to zero for consistency with `Init_TestProb_Grackle_Comoving.cpp`

YuriOku commented on example/test_problem/Hydro/Grackle_Comoving/plot/coolingrate_proper.c:
```diff
@@ -0,0 +1,356 @@
+  // Access the parameter storage with the struct you've created
+  // or with the grackle_data pointer declared in grackle.h (see further below).
+  grackle_data->use_grackle = 1;            // chemistry on
+  grackle_data->with_radiative_cooling = 0; // cooling on
```
You are right, and I have modified the comment to "cooling off".

YuriOku commented on example/test_problem/Hydro/Grackle_Comoving/plot/coolingrate_proper.c:
```diff
@@ -0,0 +1,356 @@
+      }
+      double mu = temperature[0] / (my_fields.internal_energy[0] * (gamma[0] - 1.) * temperature_units);
+      my_fields.internal_energy[0] = pow(10.0, logT) / (mu * (gamma[0] - 1.) * temperature_units);
+      my_fields.metal_density[0] = grackle_data->SolarMetalFractionByMass * my_fields.density[0];
```
It is because we only have one cell

YuriOku commented on example/test_problem/Hydro/Grackle_Comoving/plot/coolingrate_proper.c:
```diff
@@ -0,0 +1,356 @@
+        return EXIT_FAILURE;
+      }
+
+      diff = fmax(diff, fabs(temperature[0] - pow(10.0, logT)) / temperature[0]);
```
I don't think so because the cooling is turned off, and the internal energy should remain constant.

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,555 @@
+         if ( Aux_CheckFileExist(FileName) )    Aux_Message( stderr, "WARNING : file \"%s\" already exists !!\n", FileName );
+
+         FILE *File_User = fopen( FileName, "a" );
+         fprintf( File_User, "#%13s%14s%3s%14s%14s%14s%14s%14s",  "Time", "Step", "", "dt [s]", "n [1/cc]", "mu", "Temp [K]", "Edens [erg/cc], Lcool[erg cm^3 s^-1]" );
```
Fixed the typo.

YuriOku commented on example/test_problem/Hydro/Grackle_Comoving/plot/coolingrate_proper.c:
```diff
@@ -0,0 +1,356 @@
+        diff = fmax(diff, fabs(my_fields.DII_density[i] - my_fields_copy.DII_density[i]) / my_fields.DII_density[i]);
+        diff = fmax(diff, fabs(my_fields.HDI_density[i] - my_fields_copy.HDI_density[i]) / my_fields.HDI_density[i]);
+        diff = fmax(diff, fabs(my_fields.e_density[i] - my_fields_copy.e_density[i]) / my_fields.e_density[i]);
+      }
```
moved the convergence check to the end of loop

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,555 @@
+         if ( Aux_CheckFileExist(FileName) )    Aux_Message( stderr, "WARNING : file \"%s\" already exists !!\n", FileName );
+
+         FILE *File_User = fopen( FileName, "a" );
+         fprintf( File_User, "#%13s%14s%3s%14s%14s%14s%14s%14s",  "Time", "Step", "", "dt [s]", "n [1/cc]", "mu", "Temp [K]", "Edens [erg/cc], Lcool[erg cm^3 s^-1]" );
```
added a brief description for each variables

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,555 @@
+      const double Temp              = my_temperature[0];
+      const double Lcool             = Edens / fabs(my_cooling_time[0] * UNIT_T) / n / n;
+
+      fprintf( File_User, "%14.7e%14ld%3s%14.7e%14.7e%14.7e%14.7e%14.7e%14.7e", Time[0], Step, "", dt_SubStep, n, mu, Temp, Edens, Lcool );
```
I have clarified that these variables are in the physical frame in the header.

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,555 @@
+      const double n                 = Dens / CUBE(Time[0]) * UNIT_D / mu / Const_mp;
+      const double Edens             = Eint * UNIT_P / SQR(Time[0]) / CUBE(Time[0]);
+      const double Temp              = my_temperature[0];
+      const double Lcool             = Edens / fabs(my_cooling_time[0] * UNIT_T) / n / n;
```
I agree with your suggestion, and I have integrated the function in `coolingrate_proper.c` in `Aux_CoolingCurve.cpp`. The input function is called in `End_GrackleComoving`. The same input parameters are used for generating the cooling rate table, preventing potential inconsistencies.

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,555 @@
+   double Dens, MomX, MomY, MomZ, Eint, Etot;
+
+#  ifdef SUPPORT_GRACKLE
+   const double mu                = 4 / (8 - 5 * (1 - grackle_data->HydrogenFractionByMass)); // fully ionized gas
```
I mentioned it in `README`.

YuriOku commented:
Hi @hsinhaoHHuang , thank you for your review. I have updated the code and answered your comments. Following your suggestion, I have integrated the function of `coolingrate_proper.c` in the `GrackleComoving` test problem. Could you check the code again? Thanks!

hsinhaoHHuang commented on src/TestProblem/Hydro/Grackle_Comoving/Aux_CoolingCurve.cpp:
Thank you again for integrating this into the runtime calculation. I appreciate that it is simpler to run this test problem now.
About this file, I have the following comments:
- I would suggest using a function name the same as the file name. Also, adding a description comment block for this function would be great.
- Could you modify it to follow a consistent code format with GAMER? For example, three-space indentation, spaces in the `for`, `if` statements, and no trailing space.

YuriOku replied:
I have updated the code style of `Aux_CoolingCurve.cpp`

hsinhaoHHuang commented on src/TestProblem/Hydro/Grackle_Comoving/Aux_CoolingCurve.cpp:
```diff
@@ -0,0 +1,206 @@
+
+  fclose(fp);
+
+  Aux_Message( stdout, "  done %s (maxiter = %d)\n", ofname, maxiter );
```
I guess the word `done` should be put in the Line. 380 of `Init_TestProb_Grackle_Comoving.cpp`?

hsinhaoHHuang commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,610 @@
+   {
+      Aux_Message( stdout, "==========================================\n");
+      Aux_Message( stdout, "Generating cooling rate tables...\n");
+	   Compute_CoolingCurve( 0, 4, 8, 0.02);
+	   Compute_CoolingCurve( 1, 4, 8, 0.02);
+	   Compute_CoolingCurve( 2, 4, 8, 0.02);
+	   Compute_CoolingCurve( 3, 4, 8, 0.02);
+	   Compute_CoolingCurve( 4, 4, 8, 0.02);
+	   Compute_CoolingCurve( 5, 4, 8, 0.02);
+	   Compute_CoolingCurve( 6, 4, 8, 0.02);
+	   Compute_CoolingCurve( 7, 4, 8, 0.02);
+	   Compute_CoolingCurve( 8, 4, 8, 0.02);
+	   Compute_CoolingCurve( 9, 4, 8, 0.02);
```
Would you think we can support a more flexible range of redshifts? We probably assume the simulations are always from z=9 to z=0 now. However, we don't really check the input `A_INIT` and `END_T` in this test problem.

I was previously thinking we could call this function in `Aux_Record_GrackleComoving()`. Then, we can adaptively output the corresponding cooling curve at the redshift (calculated from `Time`) the user is actually running (e.g., from z=13 to z=7).

YuriOku replied:
I changed it to check the `A_INIT` and `END_T` and run `Compute_CoolingCurve` in the redshift range

hsinhaoHHuang reviewed (commented):
Hi @YuriOku,

Apologies for the delayed review, and thank you for the update—it's fantastic! The improvements will definitely enhance the user experience.

I've reviewed it again, and aside from a compilation error that needs to be fixed, the rest are just minor comments. Please take another look when you have time.

Thanks!

YuriOku commented on example/test_problem/Hydro/Grackle_Comoving/plot/Compute_CoolingCurve.py:
```diff
@@ -0,0 +1,67 @@
+# Load the data
+data = np.loadtxt('../Record__GrackleComoving', skiprows=10)
+
+time          = data[:,0] # scale factor
+dt            = data[:,2] # sec
+nden          = data[:,3] # cm^-3
+mu            = data[:,4] # mean molecular weight
+temp          = data[:,5] # K
+eden          = data[:,6] # erg/cm^3
```
changed `time` to `a_scale`

YuriOku commented on example/test_problem/Hydro/Grackle_Comoving/plot/Compute_CoolingCurve.py:
```diff
@@ -0,0 +1,67 @@
+    temp_a = (temp[idx] * (a1 - a) + temp[idx+1] * (a - a0)) / (a1 - a0)
+    L_cool_a = (L_cool[idx] * (a1 - a) + L_cool[idx+1] * (a - a0)) / (a1 - a0)
+    ax[0].scatter(temp_a, L_cool_a, s=3, label='GAMER (reconstructed) (z={})'.format(z), zorder=10)
+ax[0].scatter(temp[0], L_cool[0], s=3, label='GAMER (reconstructed) (z=9)', zorder=10)
```
changed `A_INIT` to 0.099

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Aux_CoolingCurve.cpp:
```diff
@@ -0,0 +1,206 @@
+
+      if (diff < 1.e-3)   break; // converged
+
+      iter++;
```
moved the counter before the break of loop

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,610 @@
+#  endif // #ifdef SUPPORT_GRACKLE
+
+// convert the time-step size to comoving coordinates
+   const double dt_user = dt_user_phy / SQR(Time[lv]);
+
+   return dt_user / SQR(Time[lv]);
```
Thanks for pointing it out. I have fixed it.

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
modified the code following your suggestions

YuriOku commented on src/TestProblem/Hydro/Grackle_Comoving/Aux_CoolingCurve.cpp:
```diff
@@ -0,0 +1,206 @@
+
+  my_fields.density      [0] = 1.0;
+  if ( GRACKLE_PRIMORDIAL >= GRACKLE_PRI_CHE_NSPE6 ) {
+  my_fields.HI_density   [0] = grackle_data->HydrogenFractionByMass * my_fields.density[0];
+  my_fields.HII_density  [0] = 0.0;
+  my_fields.HeI_density  [0] = (1.0 - grackle_data->HydrogenFractionByMass) * my_fields.density[0];
```
I was not careful about this because the ionization fraction will be modified to the equilibrium state. I have changed the lines to the setup in `SetGridIC()` for consistency.

hsinhaoHHuang commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -60,13 +63,16 @@ void Validate()
 #  endif
 
+   if ( !OPT__RECORD_USER )
+￼     Aux_Error( ERROR_INFO, "must turn on OPT__RECORD_USER for this test !!\n" );
```
This will cause this error during compilation:

```
TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp(67): error: unrecognized token
  ￼     Aux_Error( ERROR_INFO, "must turn on OPT__RECORD_USER for this test !!\n" );
  ^

TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp(67): error: expected an expression
  ￼     Aux_Error( ERROR_INFO, "must turn on OPT__RECORD_USER for this test !!\n" );
  ^

compilation aborted for TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp (code 2)
make: *** [Object/__cpu__Init_TestProb_Grackle_Comoving.o] Error 2
```

hsinhaoHHuang reviewed (commented):
@YuriOku
Thank you so much for the update!

I encountered a compilation error. I left the comment in `src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp`.
I do not have any other issues besides that.

YuriOku commented:
@hsinhaoHHuang Thank you for your kind suggestions. I have applied the changes and checked that the compilation works.

hsinhaoHHuang reviewed (approved):
@YuriOku
Thank you for the quick fix.

hyschive commented:
@YuriOku Thanks for the contribution! I'll do a final check before merging. Could you resolve the conflicts first?

YuriOku commented:
@hyschive I have resolved the conflicts. Thanks!

hyschive commented on example/test_problem/Hydro/Grackle_Comoving/Input__TestProb:
```diff
@@ -0,0 +1,3 @@
+# problem-specific runtime parameters
+GrackleComoving_InitialTemperature 1e8        # initial temperature (set to a 1e8K to draw a cooling function of T<1e8K)
```
```suggestion
GrackleComoving_InitialTemperature 1e8        # initial temperature (set to 1e8 K to draw a cooling function of T<1e8 K)
```

hyschive commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,637 @@
+   const double temperature_units = get_temperature_units(&Che_Units);                             // set temperature units
+
+// convert GracleComoving_InitialTemperature to comoving internal energy assuming mu above
+// --> the mean molecular weight can be different from MOLECULAR_WEIGHT
```
```suggestion
// --> the mean molecular weight mu can be different from MOLECULAR_WEIGHT
```

hyschive commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,637 @@
+   my_fields.grid_start     = new int [3];
+   my_fields.grid_end       = new int [3];
+   for (int i=0; i<3; i++) {
+      my_fields.grid_dimension[i] = 1; // the active dimension not including ghost zones.
```
For consistency, should we replace `1` with `field_size`?

hyschive commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,637 @@
+//
+// Return      :  dt
+//-------------------------------------------------------------------------------------------------------
+double Mis_GetTimeStep_GrackleComoving( const int lv, const double dTime_dt )
```
Replace `[FluSg][0]` with `[FluSg][lv]` throughout this file. This won't change the results but will improve clarity.

hyschive commented on src/TestProblem/Hydro/Grackle_Comoving/Init_TestProb_Grackle_Comoving.cpp:
```diff
@@ -0,0 +1,637 @@
+      const double n                 = Dens / CUBE(Time[0]) * UNIT_D / mu / Const_mp;                                               // Hydrogen number density
+      const double Edens             = Eint * UNIT_P / SQR(Time[0]) / CUBE(Time[0]);                                                // internal energy density
+      const double Temp              = my_temperature[0];                                                                           // temperature
+      const double Lcool             = Edens / fabs(my_cooling_time[0] * UNIT_T) / n / n;                                           // cooling rate obtained from grackle
```
Related to my comments above, IIUC, `n` actually refers to the total number density instead of the hydrogen number density. However, I thought the cooling rate is typically normalized to the square of the hydrogen number density.

I'm not very familiar with the convention used in this community. Please feel free to correct me if I'm mistaken.

hyschive commented on example/test_problem/Hydro/Grackle_Comoving/README:
```diff
@@ -0,0 +1,29 @@
+1. Default resolution = 16^3
+2. Uniform density = 1.0 (cosmic mean density)
+3. Use the cloudy cooling table for H and He (GRACKLE_PRIMORDIAL = 0)
+   --> Hydrogen's Lyman alpha cooling peak at 10^4 K disappears if non-equilibrium chemistry is used (GRACKLE_PRIMORDIAL > 0) due to the short cooling time compared to the equilibrium time scale.
```
Do I understand correctly that although this test adopts `GRACKLE_PRIMORDIAL 0` by default, it also supports `GRACKLE_PRIMORDIAL > 0`? If so, consider clarifying that.

YuriOku commented on src/Init/Init_Field.cpp:
```diff
@@ -101,6 +101,8 @@ void Init_Field()
 // 2. add other predefined fields
 #  ifdef SUPPORT_GRACKLE
+// The masses of electron and deuterium are ignored in normalization in Grackle
+// --> see the subroutine make_consistent_g in solve_rate_cool_g.F in the Grackle source code
```
Thank you for pointing this out. I meant that we should remove `Electron, DI, DII, HDI` from the normalization list. This branch originally implemented that behavior as described in the comment, but it was lost when I merged in the main branch…

YuriOku commented on src/Grackle/Grackle_AdvanceDt.cpp:
```diff
@@ -26,8 +26,15 @@
+#  ifdef COMOVING
+// convert dt from the comoving time interval to the physical time interval
+// --> see Equation (15) of Schive, Tsai, & Chiueh (2010)
+   const double dt_grackle = dt * SQR(TimeOld);
```
Thank you for the suggestion! I implemented `dt_grackle = 2.0*dt*SQR(TimeOld*TimeNew)/( SQR(TimeOld) + SQR(TimeNew) )`
